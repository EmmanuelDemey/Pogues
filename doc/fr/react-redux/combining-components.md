# Associer des composants

Voyons comment utiliser le composant simple `CodeEditor` pour construire un composant plus complexe, le `CodeListEditor`.

```javascript
import React
import CodeEditor from './code-editor'

export default function CodeListEditor() {
  return (
    <div>
      {/* 
        We won't implement the edition of the code list Label.
        Notice that comments look different from HTML in JSX.
      */}
      <button>Add a code</button>
      <div>
        <CodeEditor />
        <CodeEditor />
        <CodeEditor />
      </div>
    </div>
  )
}
```

Jusqu'à maintenant, chaque `CodeEditor` est identique, avec comme label `unhappy`. Dans la section suivante, nous verrons comment passer des paramètres à un composant pour le personnaliser.

Essayez le code avec ce [pen](http://codepen.io/BoogalooJB/pen/ggGXvp).

<!-- Add script to embed codepens -->
<script async src="https://production-assets.codepen.io/assets/embed/ei.js"></script>
<p
  data-height="700"
  data-theme-id="dark"
  data-slug-hash="ggGXvp"
  data-default-tab="js,result"
  data-user="BoogalooJB"
  data-embed-version="2"
  data-pen-title="React and Redux within Pogues"
  class="codepen" />
