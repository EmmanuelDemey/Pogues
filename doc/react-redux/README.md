# React and Redux usage in Pogues

This section is aimed at drawing a global picture of the inner workings of the application, while providing the necessary background for people who are new to React and Redux. 

Hence, you can try to follow along with no prior knowledge of these technologies, but you might need to refer at some points to some external resources to find more detailed information.

We will study the already implemented "edit a list of codes" functionality to show how the application works.

![code list editor](../img/code-list-editor.png)

