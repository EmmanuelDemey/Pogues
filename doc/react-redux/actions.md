# Actions

TODO describe how to dispatch actions