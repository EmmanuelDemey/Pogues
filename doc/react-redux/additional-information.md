# Miscellaneous informations about React and Redux

This section presents some additional information you might find useful if you are new to these technologies.

## Proptypes

TODO 
A good way to document React components.

## Design process

Adding some new components to our application can follow this process:
1. first, create some elementary static components;
2. then combine these components to build a more complex component;
3. eventually, connect the resulting component to the store if we need to pass it some data or some action creators which would not be available to its parent component. 

There is not one unique way to choose which components need to be connected to the store. In our example, we could have chosen not to connect the `CodeListEditor`, but instead, pass it all the props it needed from its parent component (the `CodeListPicker` component), which, in turn, could be connected or not to the store. Choosing the right component to connect to the store can be in some way a matter of [personal preferences](https://github.com/reactjs/redux/issues/419), but, still, [avoiding to connect too many component to the store](http://redux.js.org/docs/basics/UsageWithReact.html#presentational-and-container-components) can be a rule of thumb. 


