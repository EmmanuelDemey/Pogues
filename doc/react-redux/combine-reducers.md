# Combine reducers

TODO
combine all reducers
+ integrity checks reducers