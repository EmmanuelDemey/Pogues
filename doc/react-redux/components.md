# React components

A piece of UI in React is called a component. We will start working with a simplified version of the `CodeEditor` component from the [src/js/components/code-editor.js](https://github.com/InseeFr/Pogues/blob/master/src/js/components/code-editor.js) file:

![code editor component](../img/code-editor.png)

In order to promote design of reusable components and enforce a modular coding style, components should be as elementary as possible. Then, we can build more complex components by [composing simple ones](./combining-components.md). From this perspective, our `CodeEditor` should be seen as a composition of two different components:
- the input field;
- the controls, which in turn could be split into three components, one for each button.

Yet, to keep things simple and to stick to the code, we will consider the `CodeEditor` as an elementary block.

If we ignore layout and style instructions, `html` for this editor looks like this:

```html
<div>
  <input type="text" defaultValue="very unhappy" />
  <div class="controls">
    <button class="up"></button>
    <button class="down"></button>
    <button class="removetrash"></button>
  </div>
</div>
```

Which can be represented in React with:

```javascript
import React from 'react'

export default function CodeEditor() {
  return (
    <div>
      <input type="text" defaultValue="very unhappy" />
       <div className="controls">
          <button className="up"></button>
          <button className="down"></button>
          <button className="remove"></button>
       </div>
    </div>
  )
}
```

We notice that:
- the `CodeEditor` component is represented as a function;
- this function returns some `html` like syntax; this is called `JSX` and provides a convenient way to write components; [during the compilation process](/doc/implementation-details/build-process.md), `JSX` will be transformed to regular `javascript`, and this function will eventually return a `React` element;
- there are some small differences between `JSX` and `html` (here, the attribute `className` in place of the regular `class` attribute in `html`, and the attribute `defaultValue` instead of `value`);
- we export (read more about [exports](/doc/javascript/syntax.md#export-and-import)) the component in order to use it to build the `CodeListEditor` component.

If you want to experience with the code, you can start with this [pen](http://codepen.io/BoogalooJB/pen/PWJOEP) (for now, you can ignore the code needed to bootstrap the application, and all the code stays in one pen, so there is no `import` or `export` statements).

<!-- Add script to embed codepens -->
<script async src="https://production-assets.codepen.io/assets/embed/ei.js"></script>
<p
  data-height="434"
  data-theme-id="dark"
  data-slug-hash="PWJOEP"
  data-default-tab="js,result"
  data-user="BoogalooJB"
  data-embed-version="2"
  data-pen-title="React and Redux within Pogues"
  class="codepen" />


<!-- pen content
//bootstrap the application (you can ignore this for now)
ReactDOM.render(
  <CodeEditor />,
  document.getElementById('base')
);


function CodeEditor() {
  return (
    <div>
      <input
         type="text"
         value="very unhappy" />
       <div className="controls">
          <button className="up" />
          <button className="down" />
          <button className="remove" />
       </div>
    </div>
  )
}
-->