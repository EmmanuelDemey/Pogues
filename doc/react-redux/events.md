# Listening to events

So now, we have three `CodeEditor`, each one with a different label. Let's see what happens if we want to add a new code. We can add an event handler to the "Add a code" button:

```javascript
function CodeListEditor() {
  return (
    <div>
      <button onClick={() => console.log('trying to add a button')}>
        Add a code
      </button>
      ...
    </div>
  )
}
```

If you open a console (open the developer tools for your browser or open the pen in Code Pen and use the console button at the bottom of the screen) and you click on the "Add a code button", the message will be shown. In the next section, we will try to make something more useful happen when a user clicks the button.

Play with this [pen](http://codepen.io/BoogalooJB/pen/LxzOMo).

<!-- Add script to embed codepens -->
<script async src="https://production-assets.codepen.io/assets/embed/ei.js"></script>
<p
  data-height="434"
  data-theme-id="dark"
  data-slug-hash="LxzOMo"
  data-default-tab="js,result"
  data-user="BoogalooJB"
  data-embed-version="2"
  data-pen-title="React and Redux within Pogues"
  class="codepen" />

<!-- pen content 
function CodeEditor({ label }) {
  return (
    <div>
      <input
         type="text"
         defaultValue={label} />
       <div className="controls">
          <button className="up" />
          <button className="down" />
          <button className="remove" />
       </div>
    </div>
  )
}

function CodeListEditor() {
  const codes = [{
    id: 'code_1',
    label: 'unhappy'
  }, {
    id: 'code_2',
    label: 'happy',
  }, {
    id: 'code_3',
    label: 'very happy'
  }]
  return (
    <div>
      <button
        onClick={() => console.log('trying to add a button')}>
        Add a code
      </button>
      <div>
      {
         codes.map(({ id, label }) => 
           <CodeEditor key={id} label={label} />)
      }
      </div>
    </div>
  )
}

//bootstrap the application (you can ignore this for now)
ReactDOM.render(
  <CodeListEditor />,
  document.getElementById('base')
);
-->
