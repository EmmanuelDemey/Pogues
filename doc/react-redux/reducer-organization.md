# Reducer organization

TODO (reducers + actions)