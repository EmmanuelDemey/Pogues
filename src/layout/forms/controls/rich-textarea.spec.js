import React from 'react';
import renderer from 'react-test-renderer';
import { combineReducers, createStore } from 'redux';
import { Provider } from 'react-redux';
import { reduxForm, reducer as form } from 'redux-form';
import { Field } from 'redux-form';
import TextArea from './rich-textarea';

describe('RichTextArea', () => {
  test('should have this default template', () => {
    function MyForm() {
      return <Field component={TextArea} name="checkbox" label="Label Name" />;
    }
    const store = createStore(combineReducers({ form }), { form: {} });

    const Decorated = reduxForm({ form: 'testForm' })(MyForm);
    const tree = renderer
      .create(
        <Provider store={store}>
          <Decorated />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should have a help block', () => {
    function MyForm() {
      return <Field help component={TextArea} name="checkbox" label="Label Name" />;
    }
    const store = createStore(combineReducers({ form }), { form: {} });

    const Decorated = reduxForm({ form: 'testForm' })(MyForm);
    const tree = renderer
      .create(
        <Provider store={store}>
          <Decorated />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should have a an action toolbar', () => {
    function MyForm() {
      return <Field buttons component={TextArea} name="checkbox" label="Label Name" />;
    }
    const store = createStore(combineReducers({ form }), { form: {} });

    const Decorated = reduxForm({ form: 'testForm' })(MyForm);
    const tree = renderer
      .create(
        <Provider store={store}>
          <Decorated />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
