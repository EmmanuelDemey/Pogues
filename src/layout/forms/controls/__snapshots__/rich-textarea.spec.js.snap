// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RichTextArea should accept available suggestions props 1`] = `
<div
  className="ctrl-input"
>
  <label
    htmlFor="select-checkbox"
  >
    Label Name
    
     
  </label>
  <div>
    <textarea
      id="select-checkbox"
      name="checkbox"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      value=""
    />
  </div>
</div>
`;

exports[`RichTextArea should have a help block 1`] = `
<div
  className="ctrl-input"
>
  <label
    htmlFor="select-checkbox"
  >
    Label Name
    
     
    <span
      className="help-block"
    >
      <span
        aria-hidden="true"
        className="glyphicon glyphicon-question-sign"
      />
       
      Help
       
    </span>
  </label>
  <div>
    <textarea
      id="select-checkbox"
      name="checkbox"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      value=""
    />
  </div>
</div>
`;

exports[`RichTextArea should have this default template 1`] = `
<div
  className="ctrl-input"
>
  <label
    htmlFor="select-checkbox"
  >
    Label Name
    
     
  </label>
  <div>
    <textarea
      id="select-checkbox"
      name="checkbox"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      value=""
    />
  </div>
</div>
`;

exports[`RichTextArea should use the rte lib to get desired element 1`] = `
<div
  className="ctrl-input"
>
  <label
    htmlFor="select-checkbox"
  >
    Label Name
    
     
  </label>
  <div>
    <div
      className="root"
      style={
        Object {
          "display": "flex",
          "flexDirection": "column-reverse",
        }
      }
    >
      <div
        className="root"
        style={undefined}
      >
        <div
          className="root"
        >
          <div
            className="root"
          >
            <button
              className="root root"
              disabled={undefined}
              onClick={[Function]}
              onMouseDown={[Function]}
              title="Bold"
              type="button"
            >
              <span
                className="icon-bold"
              />
            </button>
          </div>
          <div
            className="root"
          >
            <button
              className="root root"
              disabled={undefined}
              onClick={[Function]}
              onMouseDown={[Function]}
              title="Italic"
              type="button"
            >
              <span
                className="icon-italic"
              />
            </button>
          </div>
        </div>
        <div
          className="root"
        >
          <div
            className="root"
          >
            <button
              className="root root"
              disabled={true}
              onClick={[Function]}
              onMouseDown={undefined}
              onSubmit={[Function]}
              placeholder="Insert an URL or a tooltip"
              title="Link or Tooltip"
              type="button"
            >
              <span
                className="icon-link"
              />
            </button>
          </div>
          <div
            className="root"
          >
            <button
              className="root root"
              disabled={true}
              onClick={[Function]}
              onMouseDown={[Function]}
              title="Remove Link or Tooltip"
              type="button"
            >
              <span
                className="icon-remove-link"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        className="editor"
        style={undefined}
      >
        <div
          className="DraftEditor-root"
        >
          <div
            className="DraftEditor-editorContainer"
          >
            <div
              aria-activedescendant={undefined}
              aria-autocomplete={undefined}
              aria-controls={undefined}
              aria-describedby={null}
              aria-expanded={null}
              aria-label={undefined}
              aria-multiline={undefined}
              autoCapitalize={undefined}
              autoComplete={undefined}
              autoCorrect={undefined}
              className="notranslate public-DraftEditor-content"
              contentEditable={true}
              data-testid={undefined}
              onBeforeInput={[Function]}
              onBlur={[Function]}
              onCompositionEnd={[Function]}
              onCompositionStart={[Function]}
              onCopy={[Function]}
              onCut={[Function]}
              onDragEnd={[Function]}
              onDragEnter={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              onMouseUp={[Function]}
              onPaste={[Function]}
              onSelect={[Function]}
              role="textbox"
              spellCheck={true}
              style={
                Object {
                  "outline": "none",
                  "whiteSpace": "pre-wrap",
                  "wordWrap": "break-word",
                }
              }
              suppressContentEditableWarning={true}
              tabIndex={undefined}
            >
              <div
                data-contents="true"
              >
                <div
                  className="singleline"
                  data-block={true}
                  data-editor="64eq3"
                  data-offset-key="f5cl5-0-0"
                >
                  <div
                    className="public-DraftStyleDefault-block public-DraftStyleDefault-ltr"
                    data-offset-key="f5cl5-0-0"
                  >
                    <span
                      data-offset-key="f5cl5-0-0"
                      style={Object {}}
                    >
                      <br
                        data-text="true"
                      />
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
